src/internal/m365/collection/exchange/container_resolver.go

```diff
@@ -411,3 +416,125 @@ func (cr *containerResolver) populatePaths(
+// items that don't appear in includeRankedIDs are considered to have equal
+// priority but lower priority than those in includeRankedIDs.
+//
+// includeRankedIDs is the set of containers to place at the start of the result
+// of Items in the order they should appear. IDs can either be actual
+// container IDs or well-known container IDs like "inbox".
+//
+// excludeIDs is the set of IDs that shouldn't be in the results returned by
+// Items. IDs can either be actual container IDs or well-known container IDs
+// like "inbox".
+func newRankedContainerResolver(
+	ctx context.Context,
+	base graph.ContainerResolver,
+	getter containerGetter,
+	userID string,
+	includeRankedIDs []string,
+	excludeIDs []string,
+) (*rankedContainerResolver, error) {
+	if base == nil {
+		return nil, clues.New("nil base ContainerResolver")
+	}
+
+	cr := &rankedContainerResolver{
+		resolvedInclude:   make([]string, 0, len(includeRankedIDs)),
+		resolvedExclude:   make(map[string]struct{}, len(excludeIDs)),
+		ContainerResolver: base,
+	}
+
+	for _, id := range includeRankedIDs {
+		ictx := clues.Add(ctx, "container_id", id)
+
+		c, err := getter.GetContainerByID(ctx, userID, id)
+		if err != nil {
+			return nil, clues.Wrap(err, "getting ranked container").WithClues(ictx)
+		}
+
+		gotID := ptr.Val(c.GetId())
+		if len(gotID) == 0 {
+			return nil, clues.New("received ranked include container with empty ID").
+				WithClues(ictx)
+		}
+
+		cr.resolvedInclude = append(cr.resolvedInclude, gotID)
+	}
+
+	for _, id := range excludeIDs {
+		ictx := clues.Add(ctx, "container_id", id)
+
+		c, err := getter.GetContainerByID(ctx, userID, id)
+		if err != nil {
+			return nil, clues.Wrap(err, "getting exclude container").WithClues(ictx)
+		}
+
+		gotID := ptr.Val(c.GetId())
+		if len(gotID) == 0 {
+			return nil, clues.New("received exclude container with empty ID").
+				WithClues(ictx)
+		}
+
+		cr.resolvedExclude[gotID] = struct{}{}
+	}
+
+	return cr, nil
+}
+
+func (cr *rankedContainerResolver) Items() []graph.CachedContainer {
+	found := cr.ContainerResolver.Items()
+	res := make([]graph.CachedContainer, 0, len(found))
+
+	// Add the ranked items first.
+	//
+	// TODO(ashmrtn): If we need to handle a large number of ranked items we
+	// should think about making a map of the ranked items for fast lookups later
+	// in the function.
+	for _, include := range cr.resolvedInclude {
+		if c := cr.ContainerResolver.ItemByID(include); c != nil {
+			res = append(res, c)
+		}
+	}
+
+	// Add the remaining, filtering out any of the ones we need to exclude or that
+	// we already added because they were ranked.
+	for _, c := range found {
+		if _, ok := cr.resolvedExclude[ptr.Val(c.GetId())]; ok {
+			continue
+		}
+
+		if slices.Contains(cr.resolvedInclude, ptr.Val(c.GetId())) {
+			continue
+		}
+
+		res = append(res, c)
+	}
+
+	return res
+}
+
+func (cr *rankedContainerResolver) ItemByID(id string) graph.CachedContainer {
```

Comment:
Nitpicking on a corner case. If there is a possible overlap between include & exclude lists, `Items()` code above would return the overlapping container as part of `res`. However, `ItemByID()` will return `nil`. 

Should we have the same behavior across both APIs? Another thing we can do is to catch overlaps in caller's unit or integration test.